src/supernova/circuit.rs

```diff
@@ -0,0 +1,767 @@
+      the public IO of instance ui+1. In the next iteration, because ui+1 is folded
+      into Ui+1[program_counter+1], this means that Ui+1[program_counter+1] is at least as large as Ui by the
+      properties of the folding scheme. This means that the list of running instances
+      grows in each step. To alleviate this issue, we have each F0j only produce a hash
+      of its outputs as public output. In the subsequent step, the next augmented
+      function takes as non-deterministic input a preimage to this hash." pg.16
+
+      https://eprint.iacr.org/2022/1758.pdf
+    */
+
+    // Compute the new hash H(params, i+1, program_counter, z0, z_{i+1}, U_next)
+    let mut ro = G::ROCircuit::new(
+      self.ro_consts.clone(),
+      3 // params_next, i_new, program_counter_new
+        + 2 * arity // zo, z1
+        + num_augmented_circuits * (7 + 2 * self.params.n_limbs), // #num_augmented_circuits * (7 + [X0, X1]*#num_limb)
+    );
+    ro.absorb(params_next.clone());
+    ro.absorb(i_next.clone());
+    ro.absorb(program_counter_new.clone());
+    for e in &z_0 {
+      ro.absorb(e.clone());
+    }
+    for e in &z_next {
+      ro.absorb(e.clone());
+    }
+    U_next.iter().enumerate().try_for_each(|(i, U)| {
+      U.absorb_in_ro(cs.namespace(|| format!("absorb U_new {:?}", i)), &mut ro)
+    })?;
+
+    let hash_bits = ro.squeeze(cs.namespace(|| "output hash bits"), NUM_HASH_BITS)?;
+    let hash = le_bits_to_num(cs.namespace(|| "convert hash to num"), &hash_bits)?;
+
+    // We are cycling of curve implementation, so primary/secondary will rotate hash in IO for the others to check
+    // bypass unmodified hash of other circuit as next X[0]
+    // and output the computed the computed hash as next X[1]
+    u.X1
+      .inputize(cs.namespace(|| "bypass unmodified hash of the other circuit"))?;
+    hash.inputize(cs.namespace(|| "output new hash of this circuit"))?;
+
+    Ok(())
+  }
+}
+
+#[cfg(test)]
+mod tests {
+  use super::*;
+  use crate::bellperson::{shape_cs::ShapeCS, solver::SatisfyingAssignment};
+  type PastaG1 = pasta_curves::pallas::Point;
+  type PastaG2 = pasta_curves::vesta::Point;
+
+  use crate::constants::{BN_LIMB_WIDTH, BN_N_LIMBS};
+  use crate::{
+    bellperson::r1cs::{NovaShape, NovaWitness},
+    gadgets::utils::scalar_as_base,
+    provider::poseidon::PoseidonConstantsCircuit,
+    traits::{circuit_supernova::TrivialTestCircuit, ROConstantsTrait},
+  };
+
+  // In the following we use 1 to refer to the primary, and 2 to refer to the secondary circuit
+  fn test_recursive_circuit_with<G1, G2>(
+    primary_params: CircuitParams,
+    secondary_params: CircuitParams,
+    ro_consts1: ROConstantsCircuit<G2>,
+    ro_consts2: ROConstantsCircuit<G1>,
+    num_constraints_primary: usize,
+    num_constraints_secondary: usize,
+  ) where
+    G1: Group<Base = <G2 as Group>::Scalar>,
+    G2: Group<Base = <G1 as Group>::Scalar>,
+  {
+    /*
```

Comment:
Nit: It might be preferrable to use // instead of /* ... */

Reply:
Fixed